File: code/python/02_linked_list/03_reverse_list/reverse_list.py
import unittest


class ListNode:
    def __init__(self, val: int = 0, next=None) -> None:
        self.val = val
        self.next = next


class LinkedList:
    """
    反转链表 https://leetcode.cn/problems/reverse-linked-list/
    """

    def __init__(self) -> None:
        self.head = ListNode(val=0)
        self.size = 0

    def add_at_tail(self, val: int):
        current = self.head
        while current.next != None:
            current = current.next

        node = ListNode(val=val)
        current.next = node
        self.size += 1

    def reverse1(self) -> ListNode:
        """双指针方式"""
        current = self.head.next
        pre = None
        while current != None:
            next = current.next
            current.next = pre
            pre = current
            current = next

        return pre

    def reverse2(self) -> ListNode:
        """递归方式"""
        return self._help(None, self.head)

    def _help(self, pre: ListNode, head: ListNode) -> ListNode:
        if head == None:
            return pre

        next = head.next
        head.next = pre
        return self._help(pre=head, head=next)


class TestReverseList(unittest.TestCase):
    def test_reverse1(self):
        list = LinkedList()
        list.add_at_tail(1)
        list.add_at_tail(2)
        list.add_at_tail(3)
        list.add_at_tail(4)

        res = list.reverse1()
        while res != None:
            print(res.val)
            res = res.next

    def test_reverse2(self):
        list = LinkedList()
        list.add_at_tail(1)
        list.add_at_tail(2)
        list.add_at_tail(3)
        list.add_at_tail(4)

        res = list.reverse2()
        while res.next != None:
            print(res.val)
            res = res.next


if __name__ == "__main__":
    unittest.main()

File: code/php/02_linked_list/03_reverse_list/reverse_list.php
<?php

class ListNode
{
    public function __construct(
        public int $val,
        public ListNode|NULL $next
    ) {
    }
}

/**
 * 反转链表 https://leetcode.cn/problems/reverse-linked-list/
 */
class LinkedList
{
    public ListNode $head;
    public int $size;

    public function __construct()
    {
        $this->head = new ListNode(0);
        $this->size = 0;
    }

    public function addAtTail(int $val)
    {
        $current = $this->head;
        while ($current->next != Null) {
            $current = $current->next;
        }
        $node = new ListNode($val);
        $current->next = $node;
        $this->size++;
    }

    /**
     * 双指针方法
     *
     * @return ListNode
     */
    public function reverse1(): ListNode
    {
        $current = $this->head->next;
        $pre = null;
        while ($current != NUll) {
            $next = $current->next;
            $current->next = $pre;
            $pre = $current;
            $current = $next;
        }

        return $pre;
    }

    /**
     * 递归法
     *
     * @return ListNode
     */
    public function reverse2(): ListNode
    {
        return $this->help(NULL, $this->head->next);
    }

    private function help(ListNode|NULL $pre, ListNode $head): ListNode
    {
        if ($head == NULL) {
            return $pre;
        }
        $next = $head->next;
        $head->next = $pre;
        return $this->help($head, $next);
    }
}
